From bc678760a0bd256c3f79f657f3350b31f8d031c9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Mon, 9 Dec 2024 23:48:16 +0100
Subject: [PATCH 72/86] setupapi/tests: SetupCloseFileQueue already called in
 run_queue_.

=================================================================
==1316==ERROR: AddressSanitizer: heap-use-after-free on address 0x7ec6818e3740 at pc 0x6ffff518d19f bp 0x7ffffe1fa030 sp 0x7ffffe1fa078
READ of size 4 at 0x7ec6818e3740 thread T0
0830:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff518d19e in SetupCloseFileQueue /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/queue.c:436:16
    #1 0x000140079bc3 in test_rename /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/install.c:2292:5
    #2 0x000140079bc3 in func_install /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/install.c:2581:5
    #3 0x0001400c71bc in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #4 0x0001400c71bc in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #5 0x0001400c90ef in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #6 0x6ffff80e4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #7 0x6ffff82bfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

0x7ec6818e3740 is located 0 bytes inside of 104-byte region [0x7ec6818e3740,0x7ec6818e37a8)
freed by thread T0 here:
    #0 0x6ffff88fa1a1 in free /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:71:3
    #1 0x6ffff518d16f in SetupCloseFileQueue /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/queue.c:452:5
    #2 0x000140094948 in run_queue_ /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/install.c:1443:11
    #3 0x0001400793b2 in test_rename /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/install.c:2281:5
    #4 0x0001400793b2 in func_install /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/install.c:2581:5
    #5 0x0001400c71bc in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #6 0x0001400c71bc in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #7 0x0001400c90ef in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #8 0x6ffff80e4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #9 0x6ffff82bfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

previously allocated by thread T0 here:
    #0 0x6ffff88fa3d6 in calloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:91:3
    #1 0x6ffff518ccee in SetupOpenFileQueue /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/queue.c:419:19
    #2 0x000140078e46 in test_rename /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/install.c:2269:13
    #3 0x000140078e46 in func_install /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/install.c:2581:5
    #4 0x0001400c71bc in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x0001400c71bc in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #6 0x0001400c90ef in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x6ffff80e4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #8 0x6ffff82bfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

SUMMARY: AddressSanitizer: heap-use-after-free /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/queue.c:436:16 in SetupCloseFileQueue
Shadow bytes around the buggy address:
  0x7ec6818e3480: fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa
  0x7ec6818e3500: fa fa fa fa fa fa fd fd fd fd fd fd fd fd fd fd
  0x7ec6818e3580: fd fd fd fa fa fa fa fa fa fa fa fa fd fd fd fd
  0x7ec6818e3600: fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa
  0x7ec6818e3680: fa fa fd fd fd fd fd fd fd fd fd fd fd fd fd fa
=>0x7ec6818e3700: fa fa fa fa fa fa fa fa[fd]fd fd fd fd fd fd fd
  0x7ec6818e3780: fd fd fd fd fd fa fa fa fa fa fa fa fa fa fa fa
  0x7ec6818e3800: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7ec6818e3880: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7ec6818e3900: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7ec6818e3980: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1316==ABORTING
0830:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:352858: dlls/setupapi/tests/x86_64-windows/install.ok] Fehler 1
---
 dlls/setupapi/tests/install.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/setupapi/tests/install.c b/dlls/setupapi/tests/install.c
index 9f5260359a5..61608946bff 100644
--- a/dlls/setupapi/tests/install.c
+++ b/dlls/setupapi/tests/install.c
@@ -2289,7 +2289,6 @@ static void test_rename(void)
     ok(!delete_file("b/five.txt"), "File should not exist.\n");
     ok(delete_file("b/six.txt"), "File should exist.\n");
     ok(delete_file("b/seven.txt"), "File should exist.\n");
-    SetupCloseFileQueue(queue);
 
     create_file("a/one.txt");
     create_file("a/three.txt");
@@ -2312,7 +2311,6 @@ static void test_rename(void)
     ok(!delete_file("a/four.txt"), "File should not exist.\n");
     ok(!delete_file("a/five.txt"), "File should not exist.\n");
     ok(delete_file("a/six.txt"), "File should exist.\n");
-    SetupCloseFileQueue(queue);
 
     ret = delete_file("a/");
     ok(ret, "Failed to delete directory, error %lu.\n", GetLastError());
-- 
2.47.1

